src/commands/run.rs

```diff
@@ -524,6 +589,72 @@ impl RunCommand {
     }
 }
 
+struct GuestProfiler {
+    profile: Profile,
+    process: fxprof_processed_profile::ProcessHandle,
+    thread: fxprof_processed_profile::ThreadHandle,
+    start: Instant,
+}
+
+impl GuestProfiler {
+    fn new(interval: Duration) -> Self {
+        let mut profile = Profile::new(
+            "Wasmtime",
+            std::time::SystemTime::now().into(),
+            interval.into(),
+        );
+        let process = profile.add_process("main", 0, Timestamp::from_nanos_since_reference(0));
+        let thread = profile.add_thread(process, 0, Timestamp::from_nanos_since_reference(0), true);
+        let start = Instant::now();
+        Self {
+            profile,
+            process,
+            thread,
+            start,
+        }
+    }
+
+    fn sample(&mut self, store: impl AsContext) {
+        let now = Timestamp::from_nanos_since_reference(
+            (Instant::now() - self.start).as_nanos().try_into().unwrap(),
+        );
+
+        let trace = WasmBacktrace::force_capture(store);
+        // Samply needs to see the oldest frame first, but we list the newest
+        // first, so iterate in reverse.
+        let frames = Vec::from_iter(trace.frames().iter().rev().map(|frame| {
+            let frame = if let Some(name) = frame.func_name() {
+                let idx = self.profile.intern_string(name);
+                Frame::Label(idx)
+            } else {
+                // `func_offset` should always be set because we force
+                // `generate_address_map` on if profiling is enabled.
+                let func_offset = frame.func_offset().unwrap();
+                Frame::InstructionPointer(func_offset.try_into().unwrap())
+            };
+            FrameInfo {
+                frame,
+                category_pair: CategoryHandle::OTHER.into(),
+                flags: FrameFlags::empty(),
+            }
+        }));
+
+        self.profile
+            .add_sample(self.thread, now, frames.into_iter(), CpuDelta::ZERO, 1);
```

Comment:
The interning and various allocations here make me wonder: would it be better to defer this until after execution? Should this instead buffer up a `Vec<WasmBacktrace>` which is then all serialized out at the end of execution? 

Or perhaps amortizing the cost by collecting N backtraces and then flushing them all at once perhaps.

Reply:
I want to move the `GuestProfiler` type into the `wasmtime` crate primarily because I want it to be usable by embedders, but an additional reason is because then it's reasonable for it to use `wasmtime-runtime`'s low-level stack-walking API directly. I think there's a lot of work that can be avoided that way.

I don't think batching up `WasmBacktrace` objects is going to help much, because just collecting one of those involves looking up which function each PC comes from, along with information like the function name. At that point I'd guess three hash-table lookups per function for interning aren't going to make much difference.

I'd like to merge this with the current approach, of eagerly consuming the high-level `WasmBacktrace`, as-is for now. I don't think the public API of `GuestProfiler` will need to change to support more efficient collection of backtraces, so I think we can iterate on performance improvements afterward without impacting any users.

I definitely want to improve this over time though. I think eventually we're going to want API changes in `fxprof-processed-profile` for the highest performance sampling we can do in-process.